File: app/src/main/java/com/github/willjgriff/ethereumwallet/ui/widget/validated/RxTextInputValidation.java
package com.github.willjgriff.ethereumwallet.ui.widget.validated;

import android.support.annotation.NonNull;

import java.util.List;

import io.reactivex.Observable;

/**
 * Created by williamgriffiths on 24/02/2017.
 */

public class RxTextInputValidation {

	private ValidTextInputListener mValidTextInputListener;
	private Observable<String> mTextChanged;
	private Observable<Boolean> mTextValid;
	private List<Validator> mValidators;

	public RxTextInputValidation(@NonNull ValidTextInputListener validTextInputListener,
	                             @NonNull List<Validator> validators,
	                             @NonNull Observable<CharSequence> textChangedObservable) {

		mValidTextInputListener = validTextInputListener;
		mValidators = validators;
		mTextChanged = textChangedObservable
			.map(String::valueOf)
			.distinctUntilChanged();

		setupTextChangedObservable();
	}

	public void setupTextChangedObservable() {
		// TODO: Debug these Observables and check they behave as expected (they may be firing to often)
		mTextValid = mTextChanged
			.flatMap(this::isTextValid)
			.map(validText -> validText)
			.distinctUntilChanged();

		// Hide the error when valid text is emitted
		mTextValid
			.filter(isValid -> isValid)
			.subscribe(isValid -> mValidTextInputListener.hideError());

		// Skip until the field has a valid result
		Observable<Boolean> hotValidText = mTextValid
			.publish()
			.autoConnect();
		Observable<Boolean> skipUntilValid = hotValidText
			.skipUntil(hotValidText.filter(isValid -> isValid));

		// Show the error when invalid text is emitted
		skipUntilValid
			.filter(isValid -> !isValid)
			.subscribe(isNotValid -> mValidTextInputListener.showError());
	}

	private Observable<Boolean> isTextValid(String inputText) {
		return Observable.fromIterable(mValidators)
			.all(validator -> validator.isValid(inputText))
			.toObservable();
	}

	public void setValidateTrigger(Observable<Object> validateTrigger) {
		validateTrigger
			.flatMap(anyEmission -> mTextValid)
			.filter(isValid -> !isValid)
			.subscribe(isNotValid -> mValidTextInputListener.showError());
	}

	public Observable<String> getTextChangedObservable() {
		return mTextChanged;
	}

	public Observable<Boolean> getTextValidObservable() {
		return mTextValid;
	}

	public interface ValidTextInputListener {
		void showError();
		void hideError();
	}
}

File: app/src/main/kotlin/com/github/willjgriff/ethereumwallet/ui/widget/validatedtextinput/RxTextInputValidation.kt
package com.github.willjgriff.ethereumwallet.ui.widget.validatedtextinput

import com.github.willjgriff.ethereumwallet.ui.widget.validatedtextinput.validators.Validator
import io.reactivex.Observable

/**
 * Created by williamgriffiths on 04/05/2017.
 */
class RxTextInputValidation(private val validTextInputListener: RxTextInputValidation.ValidTextInputListener,
                            private val validators: List<Validator>,
                            textChangedObservable: Observable<CharSequence>) {

    lateinit var textValidObservable: Observable<Boolean>
        private set

    val textChangedObservable: Observable<String> = textChangedObservable
            .map { it.toString() }
            .distinctUntilChanged()

    init {
        // TODO: Debug these Observables and check they behave as expected (they may be firing to often)
        setupTextValidObservable()
    }

    fun setupTextValidObservable() {
        textValidObservable = textChangedObservable
                .flatMap { isTextValid(it) }
                .map { validText -> validText }
                .distinctUntilChanged()

        // Hide the error when valid text is emitted
        textValidObservable
                .filter { isValid -> isValid }
                .subscribe { _ -> validTextInputListener.hideError() }

        // Skip until the field has a valid result
        val hotValidText = textValidObservable
                .publish()
                .autoConnect()
        val skipUntilValid = hotValidText
                .skipUntil(hotValidText.filter { isValid -> isValid })

        // Show the error when invalid text is emitted
        skipUntilValid
                .filter { isValid -> !isValid }
                .subscribe { _ -> validTextInputListener.showError() }
    }

    private fun isTextValid(inputText: String): Observable<Boolean> {
        return Observable.fromIterable(validators)
                .all { validator -> validator.isValid(inputText) }
                .toObservable()
    }

    fun setValidateTrigger(validateTrigger: Observable<Any>) {
        validateTrigger
                .flatMap { _ -> textValidObservable }
                .filter { isValid -> !isValid }
                .subscribe { _ -> validTextInputListener.showError() }
    }

    interface ValidTextInputListener {
        fun showError()
        fun hideError()
    }
}